### ISSUE TITLE
Warn on calls to core atom fns with deref-ed atom arg

### ISSUE THREAD

**Issue by coltnz** [clj-kondo/clj-kondo#955]
**Is your feature request related to a problem? Please describe.**
I keep calling swap! and reset! with @atom, I'd like a warning from clj-kondo.

**Describe the solution you'd like**
I have implemented a linter that warns me for swap!, swap-vals!, reset! and reset-vals!

**Describe alternatives you've considered**
Did consider doing more broadly for other reference types, but they're used so much less.

**Additional context**
I have impl done  https://github.com/coltnz/clj-kondo/commit/633535395f3bd7cd618548134f996961658c47df

 e.g `clojure.core/swap! called with deref-ed argument`

**Comment by borkdude** [comment]
@coltnz Have you ran `script/diff` to see if there are any false positives with this linter?

**Comment by coltnz** [comment]
Yep there are. In cljs/analyzer.cljc two atoms in atoms, first time I've seen @@ .. 

cljs/analyzer.cljc:4498:16: warning: clojure.core/swap! called with deref-ed argument
cljs/analyzer.cljc:4501:16: warning: clojure.core/swap! called with deref-ed argument

Seems OK though? Pretty edge case.

```
           @@registry-ref)})
      (when-let [speced-vars (:speced-vars spec-vars)]
......
           @@speced-vars)}))))
....
      (when registry-ref
        (swap! @registry-ref into registry)))
    (when-let [vars (seq (:cljs.spec/speced-vars cached-ns))]
      (when speced-vars
        (swap! @speced-vars into vars)))))
```

**Comment by borkdude** [comment]
So this is one edge case:

```
user=> (def f (future (atom 1)))
#'user/f
user=> (swap! @f inc)
2

user=> (def x (atom 1))
#'user/x
user=> (swap! @(resolve 'x) inc)
2
```

Deref isn't always applied to an atom, but can also be applied to futures, delays, vars.

Although not common, it would be nice if clj-kondo could do better here? Maybe it could only apply this linter when it's sure the deref happens on an atom? An alternative would be to make a configuration rule to be able to switch this off locally, maybe in the context of a specific `defn`. There's also an issue for turning off warnings on certain calls in general, see #818.

**Comment by coltnz** [comment]
Yeah I forgot about futures and delays and that will definitely lead to more false positives than I was thinking.
